packages/contracts/src/proofs.ts

```diff
@@ -0,0 +1,404 @@
+    public epochKey: BigNumberish
+    public globalStateTree: BigNumberish
+    public attesterId: BigNumberish
+    public userHasSignedUp: BigNumberish
+
+    constructor(
+        _publicSignals: BigNumberish[],
+        _proof: SnarkProof,
+        prover?: Prover
+    ) {
+        super(_publicSignals, _proof, prover)
+        this.epoch = _publicSignals[0]
+        this.epochKey = _publicSignals[1]
+        this.globalStateTree = _publicSignals[2]
+        this.attesterId = _publicSignals[3]
+        this.userHasSignedUp = _publicSignals[4]
+        this.circuit = Circuit.proveUserSignUp
+    }
+
+    public hash() {
+        const abiEncoder = utils.defaultAbiCoder.encode(
+            [
+                `tuple(
+                    uint256 epoch,
+                    uint256 epochKey,
+                    uint256 globalStateTree,
+                    uint256 attesterId,
+                    uint256 userHasSignedUp,
+                    uint256[8] proof
+                  )
+            `,
+            ],
+            [this]
+        )
+        return utils.keccak256(abiEncoder)
+    }
+}
+
+class UserTransitionProof extends BaseProof implements IUserTransitionProof {
+    public newGlobalStateTreeLeaf: BigNumberish
+    public epkNullifiers: BigNumberish[]
+    public transitionFromEpoch: BigNumberish
+    public blindedUserStates: BigNumberish[]
+    public fromGlobalStateTree: BigNumberish
+    public blindedHashChains: BigNumberish[]
+    public fromEpochTree: BigNumberish
+
+    private numEpochKeyNoncePerEpoch: number
+
+    constructor(
+        _publicSignals: BigNumberish[],
+        _proof: SnarkProof,
+        prover?: Prover,
+        numEpochKeyNoncePerEpoch = NUM_EPOCH_KEY_NONCE_PER_EPOCH
+    ) {
+        super(_publicSignals, _proof, prover)
+        this.numEpochKeyNoncePerEpoch = numEpochKeyNoncePerEpoch
+        this.newGlobalStateTreeLeaf = _publicSignals[0]
+        this.epkNullifiers = []
+        this.blindedUserStates = []
+        this.blindedHashChains = []
+        for (let i = 0; i < this.numEpochKeyNoncePerEpoch; i++) {
+            this.epkNullifiers.push(_publicSignals[1 + i])
+        }
+        this.transitionFromEpoch =
+            _publicSignals[1 + this.numEpochKeyNoncePerEpoch]
+        this.blindedUserStates.push(
+            _publicSignals[2 + this.numEpochKeyNoncePerEpoch]
+        )
+        this.blindedUserStates.push(
+            _publicSignals[3 + this.numEpochKeyNoncePerEpoch]
+        )
+        this.fromGlobalStateTree =
+            _publicSignals[4 + this.numEpochKeyNoncePerEpoch]
+        for (let i = 0; i < this.numEpochKeyNoncePerEpoch; i++) {
+            this.blindedHashChains.push(
+                _publicSignals[5 + this.numEpochKeyNoncePerEpoch + i]
+            )
+        }
+        this.fromEpochTree =
+            _publicSignals[5 + this.numEpochKeyNoncePerEpoch * 2]
+        this.circuit = Circuit.userStateTransition
+    }
+
+    public hash() {
```

Comment:
I've been thinking about this and maybe we should use getters for values that are stored and functions for values that are calculated. If we use a getter someone might assume it's just retrieving a stored value and, for example, access the hash value thousands of times in a loop, instead of storing the value locally and comparing.

So this function could stay a function, and in the `Attestation` class we instead convert that to a function too, because it's actually calculating the hash on access.